### DISCUSSION TITLE
AS4_ERROR_MESSAGE_RECEIVED

### DISCUSSION THREAD

**Discussion by MakakWasTaken** [phax/phase4#97] [Q&A] [answered]
I am getting the error: `AS4_ERROR_MESSAGE_RECEIVED`, but I am a bit unsure about how to actually get the content of the received error message. The only thing I can extract from the response of `AS4_ERROR_MESSAGE_RECEIVED` is whether it fails or not.

I have also tried looking at the `ENTSOGReceiveMessageProcessorSPI`, but it doesn't seem to be called when I am receiving the above error.

**Comment by phax** [comment, marked as answer]
The "Error Message Received" error comes only when sending messages.

So you can configure the sender builder with the following options:
1. Either use `rawResponseConsumer` with a callback of the type `IAS4RawResponseConsumer` that will give you access to the `byte[]` content of the response received or
2. use `signalMsgConsumer` with a callback of the type `IAS4SignalMessageConsumer` that will give you access to the parsed EBMS3 SignalMessage (of type `Ebms3SignalMessage`). Please consider that this method may not be called, if the received `byte[]` does not conform to the Signal Message layout

hth

**Reply by MakakWasTaken** [reply]
That did it. Is there any way to check whether the rawResponseConsumer contains an error without decoding it or do I have to use the `EbmsSignalMessage` approach for that? If so can I find an example of how to use the `signalMsgConsumer` somewhere?

**Reply by phax** [reply]
You have to use the `signalMsgConsumer` to do this - this exactly why these 2 different APIs exist.
Unfortunately there is no good code snippet that shows how to use this properly.
But it's very simple to check the contents of `Ebms3SignalMessage`:
* If it contains `Receipt` it was a success
* If it contains one or more `Error` elements it was an error
* else the layout of the SignalMessage is invalid and it indicates a programming error on the other side

**Reply by MakakWasTaken** [reply]
A little follow-up, is there any way to get the content that is being sent also (The soap envelope)?

**Reply by phax** [reply]
Yes, dumping of messages is possible using the `IAS4IncomingDumper` and `IAS4OutgoingDumper`.
You can set this globally (see the Peppol client sending examples) or on a case-by-case basis using the `outgoingDumper` and `incomingDumper` methods on the builder.

Use e.g. `AS4OutgoingDumperSingleUse` as the base class if you do it per case (same for incoming).
If you set it globally, then looking ag `AS4OutgoingDumperFileBased` as a basis for understanding or usage might help (same for incoming). You can also use these dumpers to dump to other sinks (like a DB, S3 etc.) even though I would recommend the usage of temporary files in these cases.
